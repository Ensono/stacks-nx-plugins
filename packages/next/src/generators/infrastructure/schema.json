{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "Next",
    "title": "Ensono Stacks Next Infrastructure Generator",
    "description": "The @ensono-stacks/next:infrastructure will add deployment infrastructure to an existing next project. Including generating a docker file, Heml chart and Terraform config",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        },
        "openTelemetry": {
            "type": "boolean",
            "description": "Auto-instrumentation for opentelemetry",
            "alias": "o",
            "$default": "false",
            "x-prompt": "Would you like auto-instrumentation for OpenTelemetry?"
        }
    },
    "required": ["project"]
}
