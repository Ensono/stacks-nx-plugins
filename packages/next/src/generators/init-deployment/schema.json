{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "InitDeployment",
    "title": "Ensono Stacks Next Init Deployment Generator",
    "description": "The @ensono-stacks/next:init-deployment will add deployment infrastructure to an existing next project. Including generating a docker file, Heml chart and Terraform config",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        },
        "libraryName": {
            "type": "string",
            "description": "Name provided for generated helm chart library",
            "alias": "l",
            "default": "stacks-helm-chart",
            "x-prompt": "Would you like to specify a name for the generated helm chart library?"
        }
    },
    "required": ["project"]
}
