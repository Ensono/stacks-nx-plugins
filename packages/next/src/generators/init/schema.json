{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "Next",
    "title": "",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        }
    },
    "required": ["project"]
}
