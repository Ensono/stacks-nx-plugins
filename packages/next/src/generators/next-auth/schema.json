{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "NextAuth",
    "title": "Ensono Stacks Next Auth Generator",
    "description": "The @ensono-stacks/next:next-auth generator will install and configure NextAuth.js into an existing Next application",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project?",
            "x-priority": "important"
        },
        "provider": {
            "type": "string",
            "description": "The provider to be installed.",
            "enum": ["none", "azureAd"],
            "default": "none",
            "x-prompt": "Which provider do you want to add?"
        },
        "skipPackageJson": {
            "type": "boolean",
            "default": false,
            "description": "Do not add dependencies to `package.json`."
        }
    },
    "required": ["project"]
}
