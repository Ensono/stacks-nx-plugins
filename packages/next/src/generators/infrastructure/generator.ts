import { StacksConfigError } from '@ensono-stacks/core';
import {
    joinPathFragments,
    readProjectConfiguration,
    GeneratorCallback,
    Tree,
} from '@nrwl/devkit';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';

import { NextGeneratorSchema } from './schema';
import { addInfrastructure } from './utils/add-infrastructure';
import { formatFiles } from './utils/format-files';

export default async function infrastructureGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    try {
        tasks.push(...addInfrastructure(tree, project));
    } catch (error) {
        if (error instanceof StacksConfigError) {
            console.warn(chalk.yellow`Missing Stacks configuration in nx.json`);
            console.log(
                '  Infrastructure code will not be applied now, but you can rerun this task when you update the configuration.',
            );
            console.log(
                '    nx g @ensono-stacks/next:infrastructure --project [project_name]',
            );
            return () => {};
        }

        throw error;
    }

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(project.root, 'build', 'helm', '**', '*.yaml'),
    ]);

    return runTasksInSerial(...tasks);
}
