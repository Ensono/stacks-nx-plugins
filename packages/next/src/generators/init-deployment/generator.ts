import {
    formatFiles,
    executedDependantGenerator,
    hasGeneratorExecutedForProject,
    StacksError,
} from '@ensono-stacks/core';
import {
    joinPathFragments,
    readProjectConfiguration,
    GeneratorCallback,
    Tree,
} from '@nrwl/devkit';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';

import { NextGeneratorSchema } from './schema';
import { addInfrastructure } from './utils/add-infrastructure';

export default async function initDeploymentGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    if (
        !executedDependantGenerator(
            tree,
            'WorkspaceDeployment',
            options.project,
        )
    )
        return false;
    if (
        hasGeneratorExecutedForProject(
            tree,
            options.project,
            'NextInitDeployment',
            true,
        )
    )
        return false;

    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    try {
        tasks.push(...addInfrastructure(tree, project, options));
    } catch (error) {
        if (error instanceof StacksError) {
            console.warn(chalk.yellow`Missing Stacks configuration in nx.json`);
            console.log(
                '  Infrastructure code will not be applied now, but you can rerun this task when you update the configuration.',
            );
            console.log(
                '    nx g @ensono-stacks/next:init-deployment --project [project_name]',
            );
            return () => {};
        }

        throw error;
    }

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(
            'libs',
            options.libraryName,
            'build',
            'helm',
            '**',
            '*.yaml',
        ),
    ]);

    return runTasksInSerial(...tasks);
}
