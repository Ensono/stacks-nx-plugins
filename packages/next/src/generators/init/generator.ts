import {
    formatFiles,
    formatFilesWithEslint,
    addCustomTestConfig,
    deploymentGeneratorMessage,
    hasGeneratorExecutedForProject,
    verifyPluginCanBeInstalled,
    tsMorphTree,
} from '@ensono-stacks/core';
import {
    GeneratorCallback,
    joinPathFragments,
    readProjectConfiguration,
    Tree,
    updateProjectConfiguration,
} from '@nx/devkit';
import { runTasksInSerial } from '@nx/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';
import path from 'path';

import { NextGeneratorSchema } from './schema';
import { addEslint } from './utils/eslint';
import { eslintFix } from './utils/eslint-fix';
import {
    addReactAxeConfigToApp,
    addReactAxeDependency,
} from './utils/react-axe';
import { addStorybook } from './utils/storybook';
import updateTsConfig from './utils/tsconfig';

export default async function initGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    if (hasGeneratorExecutedForProject(tree, options.project, 'NextInit'))
        return false;

    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    const update = { ...project };

    if (
        project.sourceRoot &&
        project.sourceRoot ===
            update.targets?.build?.configurations?.development?.outputPath
    ) {
        update.targets.build.configurations.development = {};
    }

    updateProjectConfiguration(tree, project.name, update);

    tasks.push(addEslint(tree, project.root), addStorybook(tree, options));

    const ciCoverageConfig = {
        ci: {
            collectCoverage: true,
            coverageReporters: ['text', 'html'],
            collectCoverageFrom: [
                './**/*.{js,jsx,ts,tsx}',
                './!**/.next/**',
                './!**/*.d.ts',
                './!**/*.config.*',
                './!**/_app.*',
            ],
            codeCoverage: true,
            ci: true,
        },
    };

    await addCustomTestConfig(
        tree,
        readProjectConfiguration(tree, options.project),
        project.name,
        ciCoverageConfig,
    );

    tasks.push(
        formatFilesWithEslint(options.project),
        addReactAxeDependency(tree),
    );

    // update tsconfig.json
    updateTsConfig(
        tree,
        project,
        path.join(project.sourceRoot, 'tsconfig.json'),
        ['next.config.js'],
    );

    // update tsconfig.spec.json
    updateTsConfig(
        tree,
        project,
        path.join(project.sourceRoot, 'tsconfig.spec.json'),
    );

    const morphTree = tsMorphTree(tree);

    console.info(chalk.green`Attempting to add configuration for react-axe`);
    addReactAxeConfigToApp(project, morphTree);
    console.info(
        chalk.green`continuing execution of next:init generator after attempting to add react-axe`,
    );
    eslintFix(project, tree);

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(project.root, 'build', 'helm', '**', '*.yaml'),
    ]);

    return runTasksInSerial(...tasks, () =>
        deploymentGeneratorMessage(
            tree,
            `nx g @ensono-stacks/next:init-deployment --project ${options.project}`,
        ),
    );
}
