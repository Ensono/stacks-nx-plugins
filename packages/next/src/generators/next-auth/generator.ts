import { tsMorphTree, formatFilesWithEslint } from '@ensono-stacks/core';
import {
    generateFiles,
    joinPathFragments,
    readProjectConfiguration,
    Tree,
    logger,
    formatFiles,
} from '@nrwl/devkit';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import path from 'path';

import { NextAuthGeneratorSchema } from './schema';
import { installDependencies } from './utils/dependencies';
import { createOrUpdateLocalEnv } from './utils/local-env';
import { addAzureAdProvider } from './utils/next-auth-provider';
import { addSessionProviderToApp } from './utils/session-provider';

export default async function nextAuthGenerator(
    tree: Tree,
    options: NextAuthGeneratorSchema,
) {
    const project = readProjectConfiguration(tree, options.project);

    if (
        !tree.exists(
            joinPathFragments(
                project.root,
                'pages',
                'api',
                'auth',
                '[...nextauth].ts',
            ),
        )
    ) {
        generateFiles(tree, path.join(__dirname, 'files'), project.root, {
            template: '',
        });
    }

    const morphTree = tsMorphTree(tree);

    addSessionProviderToApp(project, morphTree);

    if (options.provider === 'azureAd') {
        addAzureAdProvider(project, morphTree);
    }

    createOrUpdateLocalEnv(project, tree, options.provider);

    await formatFiles(tree);

    return runTasksInSerial(
        !options.skipPackageJson ? installDependencies(tree) : () => {},
        formatFilesWithEslint(options.project),
        () => {
            logger.warn(`Do not forget to update your .env.local environment variables with values.
`);
        },
    );
}
