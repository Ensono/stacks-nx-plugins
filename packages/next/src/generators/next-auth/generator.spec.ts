import { Tree, readJson, updateJson } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { applicationGenerator } from '@nrwl/next';

import generator from './generator';
import { NextAuthGeneratorSchema } from './schema';
import {
    nextAppWithProviders,
    nextAppWithDestructuredProperties,
    nextAuthEmpty,
    nextAuthWithGithub,
} from './test/fixtures';

describe('next-auth generator', () => {
    let appTree: Tree;
    const options: NextAuthGeneratorSchema = {
        project: 'next-app',
        provider: 'azureAd',
    };

    beforeEach(async () => {
        appTree = createTreeWithEmptyWorkspace();
        await applicationGenerator(appTree, {
            name: 'next-app',
            style: 'css',
            standaloneConfig: false,
        });

        updateJson(appTree, 'nx.json', nxJson => ({
            ...nxJson,
            stacks: {
                business: {
                    company: 'Amido',
                    domain: 'stacks',
                    component: 'nx',
                },
                domain: {
                    internal: 'test.com',
                    external: 'test.dev',
                },
                cloud: {
                    region: 'euw',
                    platform: 'azure',
                },
                pipeline: 'azdo',
                terraform: {
                    group: 'terraform-group',
                    storage: 'terraform-storage',
                    container: 'terraform-container',
                },
                vcs: {
                    type: 'github',
                    url: 'remote.git',
                },
            },
        }));
    });

    it('should install NextAuth without a provider', async () => {
        await generator(appTree, { ...options, provider: 'none' });

        const packageJson = readJson(appTree, 'package.json');

        expect(Object.keys(packageJson.dependencies)).toEqual(
            expect.arrayContaining(['next-auth']),
        );

        expect(appTree.exists('next-app/.env.local')).toBeTruthy();
        expect(
            appTree.exists('next-app/pages/api/auth/[...nextauth].ts'),
        ).toBeTruthy();

        const appTs = appTree.read('next-app/pages/_app.tsx');

        expect(appTs.toString()).toMatchSnapshot();
    });

    it('should configure app if there are already wrapping react providers', async () => {
        appTree.write('next-app/pages/_app.tsx', nextAppWithProviders);
        await generator(appTree, options);

        const AppTsx = appTree.read('next-app/pages/_app.tsx');

        expect(AppTsx.toString()).toMatchSnapshot();
    });

    it('should configure app if pageProps is already destructured', async () => {
        appTree.write(
            'next-app/pages/_app.tsx',
            nextAppWithDestructuredProperties,
        );
        await generator(appTree, options);

        const AppTsx = appTree.read('next-app/pages/_app.tsx');

        expect(AppTsx.toString()).toMatchSnapshot();
    });

    it('should install NextAuth with a provider', async () => {
        await generator(appTree, options);

        const nextAuthTs = appTree.read(
            'next-app/pages/api/auth/[...nextauth].ts',
        );

        expect(nextAuthTs.toString()).toMatchSnapshot();
    });

    it('should safely run on an existing NextAuth install', async () => {
        await generator(appTree, options);
        await generator(appTree, options);

        const nextAuthTs = appTree.read(
            'next-app/pages/api/auth/[...nextauth].ts',
        );

        expect(nextAuthTs.toString()).toMatchSnapshot();
    });

    it('should error if an existing NextAuth install is not valid', async () => {
        appTree.write('next-app/pages/api/auth/[...nextauth].ts', '');
        await expect(generator(appTree, options)).rejects.toThrowError(
            'Unable to find the NextAuth implementation function.',
        );
    });

    it('should run on an existing NextAuth install with no providers', async () => {
        appTree.write(
            'next-app/pages/api/auth/[...nextauth].ts',
            nextAuthEmpty,
        );
        await generator(appTree, options);

        const nextAuthTs = appTree.read(
            'next-app/pages/api/auth/[...nextauth].ts',
        );

        expect(nextAuthTs.toString()).toMatchSnapshot();
    });

    it('should appendto an existing list of providers', async () => {
        appTree.write(
            'next-app/pages/api/auth/[...nextauth].ts',
            nextAuthWithGithub,
        );
        await generator(appTree, options);

        const nextAuthTs = appTree.read(
            'next-app/pages/api/auth/[...nextauth].ts',
        );

        expect(nextAuthTs.toString()).toMatchSnapshot();
    });

    it('should append to an existing .env.local', async () => {
        appTree.write('next-app/.env.local', 'NEXTAUTH_URL=http://website.com');
        await generator(appTree, options);
        const localEnv = appTree.read('next-app/.env.local');
        expect(localEnv.toString()).toContain(
            'NEXTAUTH_URL=http://website.com',
        );
        expect(localEnv.toString()).not.toContain(
            'NEXTAUTH_URL=http://localhost:4200',
        );
    });

    it('should generate redis adapter lib', async () => {
        await generator(appTree, {
            ...options,
            redisAdapter: true,
            redisEnvVar: 'REDIS_URL',
        });

        const nextAuthTs = appTree.read(
            'next-app/pages/api/auth/[...nextauth].ts',
        );
        expect(nextAuthTs.toString()).toMatchSnapshot();

        expect(appTree.exists('next-auth-redis/src/index.ts')).toBeTruthy();

        const localEnv = appTree.read('next-app/.env.local');
        expect(localEnv.toString()).toContain('REDIS_URL=localhost:6379');

        const packageJson = readJson(appTree, 'package.json');
        expect(Object.keys(packageJson.dependencies)).toEqual(
            expect.arrayContaining(['ioredis', 'uuid']),
        );
        expect(Object.keys(packageJson.devDependencies)).toEqual(
            expect.arrayContaining(['@types/uuid']),
        );
    });

    it('should generate redis adapter lib with custom env var name', async () => {
        await generator(appTree, {
            ...options,
            redisAdapter: true,
            redisEnvVar: 'REDIS_CONNECTION_STRING',
        });

        const nextAuthTs = appTree.read(
            'next-app/pages/api/auth/[...nextauth].ts',
        );
        expect(nextAuthTs.toString()).toMatchSnapshot();

        const localEnv = appTree.read('next-app/.env.local');
        expect(localEnv.toString()).toContain(
            'REDIS_CONNECTION_STRING=localhost:6379',
        );

        const redisAdapterTest = appTree.read(
            'next-auth-redis/src/index.test.ts',
        );
        expect(redisAdapterTest.toString()).toContain(
            'new Redis(process.env.REDIS_CONNECTION_STRING)',
        );
    });

    it('should generate redis adapter lib with custom name', async () => {
        await generator(appTree, {
            ...options,
            redisAdapter: true,
            redisEnvVar: 'REDIS_URL',
            redisAdapterName: 'redis-adapter-for-next-auth',
        });

        const nextAuthTs = appTree.read(
            'next-app/pages/api/auth/[...nextauth].ts',
        );
        expect(nextAuthTs.toString()).toMatchSnapshot();

        expect(
            appTree.exists('redis-adapter-for-next-auth/src/index.ts'),
        ).toBeTruthy();
    });

    it('should skip installing depndencies', async () => {
        await generator(appTree, { ...options, skipPackageJson: true });

        const packageJson = readJson(appTree, 'package.json');

        expect(Object.keys(packageJson.dependencies)).not.toEqual(
            expect.arrayContaining(['next-auth']),
        );
    });
});
