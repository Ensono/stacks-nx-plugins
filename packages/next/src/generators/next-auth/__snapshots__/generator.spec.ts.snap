// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`next-auth generator should appendto an existing list of providers 1`] = `
"import NextAuth from 'next-auth';
import GitHub from 'next-auth/providers/github';
import AzureADProvider from 'next-auth/providers/azure-ad';

export const { handlers, signIn, signOut, auth } = NextAuth({
    providers: [
        GitHub,
        AzureADProvider({
            clientId: process.env.AZURE_AD_CLIENT_ID as string,
            clientSecret: process.env.AZURE_AD_CLIENT_SECRET as string,
            tenantId: process.env.AZURE_AD_TENANT_ID as string,
        }),
    ],
});
"
`;

exports[`next-auth generator should configure app if there are already wrapping react providers 1`] = `
"function SomeProvider({ children }: PropsWithChildren) {
    return <>{children}</>;

function RootLayout({
    children,
  }: {
    children: React.ReactNode;
  }) {
    return (
      <html lang="en">
        <body>
        <SomeProvider>
          {children}
          </SomeProvider>
        </body>
      </html>
    );
  }
  
  export default RootLayout;
  "
`;

exports[`next-auth generator should install NextAuth with a provider 1`] = `
"import NextAuth from 'next-auth';
import AzureADProvider from 'next-auth/providers/azure-ad';

export const { handlers, signIn, signOut, auth } = NextAuth({
    providers: [
        AzureADProvider({
            clientId: process.env.AZURE_AD_CLIENT_ID as string,
            clientSecret: process.env.AZURE_AD_CLIENT_SECRET as string,
            tenantId: process.env.AZURE_AD_TENANT_ID as string,
        }),
    ],
});
"
`;

exports[`next-auth generator should install NextAuth without a provider 1`] = `
"import './global.css';

export const metadata = {
    title: 'Welcome to next-app',
    description: 'Generated by create-nx-workspace',
};

export default function RootLayout({
    children,
}: {
    children: React.ReactNode;
}) {
    return (
        <html lang="en">
            <body>{children}</body>
        </html>
    );
}
"
`;

exports[`next-auth generator should run on an existing NextAuth install with no providers 1`] = `
"import NextAuth from 'next-auth';

export const { handlers, signIn, signOut, auth } = NextAuth({});
"
`;

exports[`next-auth generator should safely run on an existing NextAuth install 1`] = `
"import NextAuth from 'next-auth';
import AzureADProvider from 'next-auth/providers/azure-ad';

export const { handlers, signIn, signOut, auth } = NextAuth({
    providers: [
        AzureADProvider({
            clientId: process.env.AZURE_AD_CLIENT_ID as string,
            clientSecret: process.env.AZURE_AD_CLIENT_SECRET as string,
            tenantId: process.env.AZURE_AD_TENANT_ID as string,
        }),
    ],
});
"
`;
